oak_crypto/src/noise_nk/mod.rs

```diff
@@ -0,0 +1,267 @@
+    let ephemeral_pub_key_bytes = ephemeral_priv.compute_public_key();
+    noise.mix_hash(ephemeral_pub_key_bytes.as_slice());
+    noise.mix_key(ephemeral_pub_key_bytes.as_slice());
+    let ee_ecdh_bytes = rustcrypto::p256_scalar_mult(&ephemeral_priv, &peer_pub)
+        .map_err(|_| Error::InvalidHandshake)?;
+    noise.mix_key(ee_ecdh_bytes.as_slice());
+
+    let response_ciphertext = noise.encrypt_and_hash(&[]);
+
+    let keys = noise.traffic_keys();
+    Ok(Response {
+        crypter: Crypter::new(&keys.0, &keys.1),
+        handshake_hash: noise.handshake_hash(),
+        response: [ephemeral_pub_key_bytes.as_slice(), &response_ciphertext].concat(),
+    })
+}
+
+pub mod test_client {
+    use super::*;
+
+    pub struct HandshakeInitiator {
+        noise: Noise,
+        identity_pub_key: [u8; P256_X962_LENGTH],
+        ephemeral_priv_key: P256Scalar,
+    }
+
+    impl HandshakeInitiator {
+        pub fn new(peer_public_key: &[u8; P256_X962_LENGTH]) -> Self {
+            Self {
+                noise: Noise::new(HandshakeType::Nk),
+                identity_pub_key: *peer_public_key,
+                ephemeral_priv_key: P256Scalar::generate(),
+            }
+        }
+
+        pub fn build_initial_message(&mut self) -> Vec<u8> {
+            self.noise.mix_hash(&[0; 1]);
+            self.noise.mix_hash_point(self.identity_pub_key.as_slice());
+            let ephemeral_pub_key = self.ephemeral_priv_key.compute_public_key();
+            let ephemeral_pub_key_bytes = ephemeral_pub_key.as_ref();
+
+            self.noise.mix_hash(ephemeral_pub_key_bytes);
+            self.noise.mix_key(ephemeral_pub_key_bytes);
+            let es_ecdh_bytes =
+                rustcrypto::p256_scalar_mult(&self.ephemeral_priv_key, &self.identity_pub_key)
+                    .unwrap();
+            self.noise.mix_key(&es_ecdh_bytes);
+
+            let ciphertext = self.noise.encrypt_and_hash(&[]);
+            [ephemeral_pub_key_bytes, &ciphertext].concat()
+        }
+
+        pub fn process_response(&mut self, handshake_response: &[u8]) -> ([u8; 32], Crypter) {
+            let peer_public_key_bytes = &handshake_response[..P256_X962_LENGTH];
+            let ciphertext = &handshake_response[P256_X962_LENGTH..];
+
+            let ee_ecdh_bytes = rustcrypto::p256_scalar_mult(
+                &self.ephemeral_priv_key,
+                peer_public_key_bytes.try_into().unwrap(),
+            )
+            .unwrap();
+            self.noise.mix_hash(peer_public_key_bytes);
+            self.noise.mix_key(peer_public_key_bytes);
+            self.noise.mix_key(&ee_ecdh_bytes);
+
+            let plaintext = self.noise.decrypt_and_hash(ciphertext).unwrap();
+            assert_eq!(plaintext.len(), 0);
+            let (write_key, read_key) = self.noise.traffic_keys();
+            (self.noise.handshake_hash(), Crypter::new(&read_key, &write_key))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
nit: I think we could move tests to a separate `tests.rs` file

Reply:
Done.